src/main/java/org/terasology/genome/breed/DiscreteBreedingAlgorithm.java

```diff
@@ -0,0 +1,106 @@
+import org.terasology.utilities.random.FastRandom;
+
+import java.util.Arrays;
+
+/**
+ * A breeding algorithm that produces a diploid cross for Discrete traits with a Punnet Square.
+ */
+public class DiscreteBreedingAlgorithm implements BreedingAlgorithm {
+    private float mutationChance;
+    private GeneMutator geneMutator;
+
+    public DiscreteBreedingAlgorithm(float mutationChance, GeneMutator geneMutator) {
+        this.mutationChance = mutationChance;
+        this.geneMutator = geneMutator;
+    }
+
+    /**
+     * Check whether two organisms with the given genes can breed.
+     *
+     * @param genes1 The genes of the first organism
+     * @param genes2 The genes of the second organism
+     * @return Whether the two organisms can breed
+     */
+    @Override
+    public boolean canCross(String genes1, String genes2) {
+        validateGenes(genes1, genes2);
+
+        return (genes1.compareToIgnoreCase(genes2) == 0 ? true : false);
+    }
+
+    /**
+     * Produces the genes of an offspring from parent organisms with the genes specified.
+     *
+     * @param genes1 The genes of the first parent organism
+     * @param genes2 The genes of the second parent organism
+     * @return The genes of an offspring from the two parent organisms
+     */
+    @Override
+    public String produceCross(String genes1, String genes2) {
+        if (!canCross(genes1, genes2)) {
+            return null;
+        }
+
+        int genomeLength = genes1.length();
+        int punnetCounter = 0;
+
+        FastRandom rand = new FastRandom();
+
+        StringBuilder result = new StringBuilder(genomeLength);
+
+        char[] chars1 = genes1.toCharArray();
+        char[] chars2 = genes2.toCharArray();
+        char[] resultGenes;
+
+        String[] punnetSquare = new String[4];
+        String alleles = "";
+        String resultant;
+
+        for (int i = 0; i < genomeLength; i++) {
+            alleles = "" + chars1[i];
+            for (int j = 0; j < genomeLength; j++) {
+                alleles = alleles + "" + chars2[j];
+                punnetSquare[punnetCounter++] = alleles;
+                alleles = "" + alleles.charAt(0);
+            }
+        }
+
+        resultGenes = punnetSquare[rand.nextInt(3)].toCharArray();
+        Arrays.sort(resultGenes);
+        resultant = new String(resultGenes);
+
+        result.append(resultant);
+
+        if (mutationChance >= rand.nextFloat()) {
+            int geneIndex = rand.nextInt(genomeLength);
+            char gene = geneMutator.mutateGene(rand.nextFloat(), geneIndex, result.charAt(geneIndex));
+            result.replace(geneIndex, geneIndex + 1, "" + gene);
+        }
+
+        return result.toString();
+    }
+
+    private void validateGenes(String genes1, String genes2) {
+        if (genes1 == null || genes2 == null || genes1.length() != genes2.length()) {
+            throw new IllegalArgumentException("Genomes not defined or of incorrect length");
```

Comment:
Throwing exception can potentially lead to a crash during runtime. 

Alternatives can be displaying and logging the error.

Reply:
replaced with a log and changed validateGenes to replace a boolean value